src/inmanta/logging.py

```diff
@@ -0,0 +1,245 @@
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import argparse
+import logging
+import os
+import sys
+from typing import Optional, TextIO
+
+import colorlog
+from colorlog.formatter import LogColors
+
+from inmanta import const
+
+
+def _is_on_tty() -> bool:
+    return (hasattr(sys.stdout, "isatty") and sys.stdout.isatty()) or const.ENVIRON_FORCE_TTY in os.environ
+
+
+"""
+This dictionary maps the Inmanta log levels to the corresponding Python log levels
+"""
+log_levels = {
+    "0": logging.ERROR,
+    "1": logging.WARNING,
+    "2": logging.INFO,
+    "3": logging.DEBUG,
+    "4": 2,
+    "ERROR": logging.ERROR,
+    "WARNING": logging.WARNING,
+    "INFO": logging.INFO,
+    "DEBUG": logging.DEBUG,
+    "TRACE": 2,
+}
+
+
+class InmantaLogs:
+    """
+    A class that provides logging functionality for Inmanta projects.
+
+    Usage:
+    To use this class, you first need to call the `setup_handler` method to configure the logging handler. This method
+    takes a `stream` argument that specifies where the log messages should be sent to. If no `stream` is provided,
+    the log messages will be sent to standard output.
+
+    You can then call the `apply_options` method to configure the logging options. This method takes an `options`
+    argument that should be an object with the following attributes:
+    - `log_file`: if this attribute is set, the logs will be written to the specified file instead of the stream
+      specified in `setup_handler`.
+    - `log_file_level`: the logging level for the file handler (if `log_file` is set).
+    - `verbose`: the verbosity level of the log messages.
+    - 'timed': if true,  adds the time to the formatter in the log lines.
+
+    The setup is not done in one step as we want logs for the cmd_parser, which will provide the options needed to configure
+    the 'final' logger with apply_options.
+    """
+
+    _handler: Optional[logging.Handler] = None
+
+    @classmethod
+    def create_default_handler(cls, stream: TextIO = sys.stdout) -> None:
+        """
+        Set up the logging handler for Inmanta.
+
+        :param stream: The stream to send log messages to. Default is standard output (sys.stdout).
+        """
+        cls._handler = logging.StreamHandler(stream=stream)
+        cls.set_log_level("INFO")
+        formatter = cls._get_log_formatter_for_stream_handler(timed=False)
+        cls.set_log_formatter(formatter)
+
+        logging.root.handlers = []
+        logging.root.addHandler(cls._handler)
+        logging.root.setLevel(0)
+
+    @classmethod
+    def apply_options(cls, options: object) -> None:
```

Comment:
@sanderr I don't know how to type options here. Could you help?

Reply:
Yes but it'll have to be for tomorrow I'm afraid, or perhaps even Friday.

Reply:
No issue :) Thanks!

Reply:
Why not make options into a separate object?